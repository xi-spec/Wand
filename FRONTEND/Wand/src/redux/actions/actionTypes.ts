export default {
  LOAD_PRODUCT_TYPE: 'LOAD_PRODUCT_TYPE',
  LOAD_ALL_PRODUCTS: 'LOAD_ALL_PRODUCTS',
  LOAD_NEW_PRODUCTS: 'LOAD_NEW_PRODUCTS',
  LOAD_PRODUCTS_LIST: 'LOAD_PRODUCTS_LIST',
  LOAD_PRODUCT_DETAIL: 'LOAD_PRODUCT_DETAIL',
  ADD_PRODUCT_TO_CART: ' ADD_PRODUCT_TO_CART',
  INCREASE_QUANTITY_CART: 'INCREASE_QUANTITY_CART',
  DECREASE_QUANTITY_CART: 'DECREASE_QUANTITY_CART',
  DELETE_PRODUCT_CART: 'DELETE_PRODUCT_CART',
  DELETE_ALL_CART: 'DELETE_ALL_CART',
  UPDATE_STOCK_API: 'UPDATE_STOCK_API',
  SEARCH_PRODUCT: 'SEARCH_PRODUCT',
  ADD_PRODUCT_TO_FAVOURITE: 'ADD_PRODUCT_TO_FAVOURITE',
  REGISTER: 'REGISTER',
  REGISTER_ERROR: 'REGISTER_ERROR',
  LOGIN: 'LOGIN',
  LOGIN_ERROR: 'LOGIN_ERROR',
  CLEAN_ISEXIST_STATUS: 'CLEAN_STATUS',
  CLEAN_LOGIN_ERROR: 'CLEAN_LOGIN_ERROR',
  CLEAN_REGISTED: 'CLEAN_REGISTED',
  REMOVE_FAVOURITE: 'REMOVE_FAVOURITE',
  LOGOUT: 'LOGOUT',
  UPDATE_USER: 'UPDATE_USER',
  PAYMENT_SUCCSSFULY: 'PAYMENT_SUCCESSFULY',
  PDATE_STOCK_API: 'PDATE_STOCK_API'
}
;
