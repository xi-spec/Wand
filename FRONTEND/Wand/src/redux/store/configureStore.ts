import reduxImmutableStatInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import { applyMiddleware, compose, createStore } from 'redux';
import rootReducer from '../reducers';

function configureStore () {
  const composeEnhancers = compose;

  return createStore(
    rootReducer,
    composeEnhancers(applyMiddleware(reduxImmutableStatInvariant(), thunk))
  );
}

export default configureStore();
