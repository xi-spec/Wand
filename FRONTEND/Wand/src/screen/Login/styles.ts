
import {
  StyleSheet

} from 'react-native';
import {
  COLORS,
  SIZE
} from '../../constants/theme';

const styles = StyleSheet.create({
  container: {
    marginLeft: 50,
    marginRight: 50,
    paddingTop: 50
  },

  blueBigText: {
    color: COLORS.primary,
    fontSize: 36,
    marginBottom: 5

  },
  greySmallText: {
    color: COLORS.darkGrey,
    fontSize: SIZE.primary
  },
  inputContainer: {
    paddingTop: 50
  },
  input: {
    width: 300,
    height: 50,
    fontSize: SIZE.primary,
    color: COLORS.primary,
    borderBottomWidth: 1,
    borderBottomColor: COLORS.darkGrey,
    marginBottom: 30,
    marginTop: 20

  },
  buttonContainer: {
    alignItems: 'center',
    marginTop: 70

  },
  button: {
    width: 250,
    height: 50,
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: SIZE.primary,
    backgroundColor: COLORS.primary

  },

  buttonText: {
    color: 'white',
    fontSize: SIZE.buttonSize
  },

  lineContainer: {
    paddingTop: 20,
    flexDirection: 'row'
  },
  line: {
    width: 150,
    borderBottomWidth: 1,
    borderBottomColor: COLORS.lightGrey,
    marginLeft: 5,
    marginRight: 5

  },
  orText: {
    color: COLORS.darkGrey,
    top: 7
  },

  centerContainer: {
    alignItems: 'center',
    paddingTop: 50
  },
  apple: {
    width: 50,
    height: 50
  }
});

export default styles
;
