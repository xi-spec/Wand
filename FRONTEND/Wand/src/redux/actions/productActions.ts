import actionTypes from './actionTypes';
import axios from 'axios';
import SERVER from '../../constants/server';
import { Dispatch } from 'redux';
import { Product, ShoppingCart } from '../../models/initiaState';

export function loadProductsType () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_TYPE);
    dispatch({
      type: actionTypes.LOAD_PRODUCT_TYPE,
      data
    });
  };
}

export function loadAllProducts () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_PRODUCT);
    dispatch({
      type: actionTypes.LOAD_ALL_PRODUCTS,
      data
    });
  };
}

export function loadProductsList (typeName:string) {
  return {
    type: actionTypes.LOAD_PRODUCTS_LIST,
    typeName
  };
}

export function loadProductDetail (productId:string) {
  return {
    type: actionTypes.LOAD_PRODUCT_DETAIL,
    productId
  };
}

export function addProductToCart (product:Product) {
  return {
    type: actionTypes.ADD_PRODUCT_TO_CART,
    product
  };
}

export function increaseQuantityCart (productId:string) {
  return {
    type: actionTypes.INCREASE_QUANTITY_CART,
    productId
  };
}

export function decreaseQuantityCart (productId:string) {
  return {
    type: actionTypes.DECREASE_QUANTITY_CART,
    productId
  };
}

export function deleteProductCart (productId:string, quantity:number) {
  return {
    type: actionTypes.DELETE_PRODUCT_CART,
    productId,
    quantity
  };
}

export function searchProducts (name:string) {
  return {
    type: actionTypes.SEARCH_PRODUCT,
    name
  };
}

export function updateStockApi (cart:ShoppingCart) {
  return (dispatch : Dispatch) => {
    cart.cartList.forEach(async (product) => {
      const { data } = await axios.put(`${SERVER.URL_PRODUCT}/${product._id}`, { stock: product.stock });
      dispatch({
        type: actionTypes.UPDATE_STOCK_API,
        data
      });
    });
  };
}
