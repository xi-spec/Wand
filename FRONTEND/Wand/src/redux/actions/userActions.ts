import axios from 'axios';
import actionTypes from './actionTypes';
import SERVER from '../../constants/server';
import { User, Product, ShoppingCart } from '../../models/initiaState';
import { Dispatch } from 'redux';

export function addProductToFavourite (product:Product, user:User, favouriteState:boolean) {
  let newUser;
  favouriteState
    ? newUser = { ...user, favourite: [...user.favourite, product] }
    : newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.ADD_PRODUCT_TO_FAVOURITE,
      data
    });
  };
}

export function register (email:string, name:string, password:string) {
  const query = { email, name, password };
  let actionObj;
  return async (dispatch:Dispatch) => {
    try {
      const { data, status } = await axios.post(SERVER.URL_REGISTER, query);
      actionObj = {
        type: actionTypes.REGISTER,
        data,
        status
      };
    } catch (error) {
      actionObj = {
        type: actionTypes.REGISTER_ERROR
      };
    }
    dispatch(
      actionObj
    );
  };
}

export function login (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    let actionObj;
    try {
      const { data } = await axios.post(SERVER.URL_LOGIN, query);
      actionObj = {
        type: actionTypes.LOGIN,
        data

      };
    } catch (error) {
      actionObj = {
        type: actionTypes.LOGIN_ERROR
      };
    }

    dispatch(actionObj);
  };
}

export function logout (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.post(SERVER.URL_LOGOUT, query);

    dispatch({
      type: actionTypes.LOGOUT,
      data
    });
  };
}

export function cleanIsExistStatus () {
  return {
    type: actionTypes.CLEAN_ISEXIST_STATUS
  };
}

export function cleanLoginError () {
  return {
    type: actionTypes.CLEAN_LOGIN_ERROR
  };
}

export function cleanRegisted () {
  return {
    type: actionTypes.CLEAN_REGISTED
  };
}

export function removeFavourite (product:Product, user:User) {
  const newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.REMOVE_FAVOURITE,
      data
    });
  };
}

export function updateAddress (street:string,
  city:string, country:string, postcode:string, user:User) {
  const query = { street, city, country, postcode };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { address: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function updateCard (cardNumber:string, securtyCode:string, cardholder:string, expiryDate:string, user:User) {
  const query = { cardNumber, securtyCode, cardholder, expiryDate };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { card: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function paymentSuccessfuly (user:User, cart:ShoppingCart) {
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { orderHistory: [...user.orderHistory, cart] });
    dispatch({
      type: actionTypes.PAYMENT_SUCCSSFULY,
      data
    });
  };
}
