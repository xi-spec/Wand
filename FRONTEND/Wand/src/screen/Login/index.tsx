import React, { useState, useEffect } from 'react';
import {
  Text,
  TouchableOpacity,
  View,
  TextInput,
  Image,
  ScrollView,
  Alert
} from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { login, cleanLoginError } from '../../redux/actions/userActions';
import { IMAGES } from '../../constants/images';
import { COLORS } from '../../constants/theme';
import styles from './styles';
import { User } from '../../models/initiaState';

function Login ({ actions, user, navigation }:
  {actions:any,
   navigation:{navigate:Function, reset:Function},
    user:User}) {
  const [email, setEmail] = useState(user.email || '');
  const [password, setPassword] = useState('');

  useEffect(() => {
    if (user.isLogged) {
      navigation.reset({
        index: 0,
        routes: [{ name: 'Home' }]
      });
    }
  }, [user.isLogged]);
  return (

  <ScrollView style={styles.container}>
    { user.loginError
      ? Alert.alert('', 'email or password is wrong!', [
        {
          text: 'OK',
          onPress: () => actions.cleanLoginError()
        }
      ])
      : null}
    <View >
    <Text style={styles.blueBigText}>
    Welcome!
   </Text>

   <Text style={styles.greySmallText}>
   Sign in to continue
   </Text>
    </View>

    <View style={styles.inputContainer}>
        <TextInput style={styles.input}
         testID='emailInput'
        placeholder="Email"
        placeholderTextColor={COLORS.darkGrey}
        value={email}
        onChangeText={text => setEmail(text.toLowerCase())}
        autoCapitalize='none'
        >

        </TextInput>
        <TextInput style={styles.input}
        placeholder="Password"
        placeholderTextColor={COLORS.darkGrey}
        secureTextEntry={true}
        value={password}
        testID='passwordInput'
        onChangeText={text => setPassword(text)}
        >

        </TextInput>
    </View>

   <View style={styles.buttonContainer}>
  <TouchableOpacity
   testID='login'
    style={styles.button}
    onPress={() => actions.login(email, password)}
      >
        <Text style={styles.buttonText}>LOGIN</Text>
      </TouchableOpacity>
   </View>

  <View style={styles.lineContainer}>
    <TextInput style={styles.line}/>
    <Text style={styles.orText}>o</Text>

    <TextInput style={styles.line}/>
  </View>

<TouchableOpacity style={styles.centerContainer}>
<Image source={{ uri: IMAGES.apple }} style={styles.apple}></Image>
</TouchableOpacity>

</ScrollView>
  );
}

function mapStateToProps ({ user }) {
  return { user };
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      login,
      cleanLoginError
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
