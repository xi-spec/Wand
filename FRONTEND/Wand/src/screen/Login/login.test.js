import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/userActions';
import Login from '../Login';
import thunk from 'redux-thunk';
import { Alert } from 'react-native';

describe('Given a Login component', () => {
  const mockStore = configureStore([thunk]);
  let store;
  let navigation;

  beforeEach(() => {
    jest.spyOn(actions, 'login').mockReturnValueOnce({ type: '' });
    jest.spyOn(actions, 'cleanLoginError').mockReturnValueOnce({ type: '' });
    store = mockStore({
      user: {
        email: 'ho@la.com',
        password: '123123123',
        isLogged: false
      }
    });
    navigation = {
      navigate: jest.fn(),
      reset: jest.fn()
    };
  });
  describe('When is rendered', () => {
    describe('And user.isLogged is false', () => {
      test('Then is matched to snapshot', () => {
        const tree = render(<Provider store={store}><Login/></Provider>);

        expect(tree).toMatchSnapshot();
      });
    });
    describe('And user.isLogged is true', () => {
      test('Then navigantion.reset is invoked ', () => {
        store = mockStore({
          user: {
            isLogged: true
          }
        });

        render(<Provider store={store}><Login navigation={navigation}/></Provider>);

        expect(navigation.reset).toHaveBeenCalled();
      });
    });
  });

  describe('When emailInput is typing with HELLO', () => {
    test('Then emailInput value is HELLO in lowerCase', () => {
      const { getByTestId } = render(<Provider store={store}><Login/></Provider>);
      const emailInput = getByTestId('emailInput');
      const text = 'HELLO';
      fireEvent.changeText(emailInput, text);

      expect(emailInput.props.value).toBe(text.toLowerCase());
    });
  });

  describe('When passwordInput is typing with HELLO', () => {
    test('Then passwordInput value is HELLO', () => {
      const { getByTestId } = render(<Provider store={store}><Login/></Provider>);
      const passwordInput = getByTestId('passwordInput');
      const text = 'HELLO';
      fireEvent.changeText(passwordInput, text);

      expect(passwordInput.props.value).toBe(text);
    });
  });

  describe('When login button is pressed', () => {
    test('Then navigation.navigate is invoked', () => {
      const { getByTestId } = render(<Provider store={store}><Login navigation={navigation}/></Provider>);
      const login = getByTestId('login');
      fireEvent.press(login);

      expect(actions.login).toHaveBeenCalled();
    });
  });

  describe('When user.loginError is true', () => {
    test('Then Alert.alert is invoked with a error message', () => {
      store = mockStore({
        user: {
          loginError: true
        }
      });
      jest.spyOn(Alert, 'alert').mockReturnValueOnce('error');
      render(<Provider store={store}><Login navigation={navigation}/></Provider>);

      expect(Alert.alert).toHaveBeenCalled();
    });
    describe('And pressed ok ', () => {
      test('Then actions.cleanLoginError is invoked', () => {
        store = mockStore({
          user: {
            loginError: true
          }
        });
        render(<Provider store={store}><Login navigation={navigation}/></Provider>);
        jest.spyOn(Alert, 'alert').mock.calls[0][2][0].onPress();

        expect(actions.cleanLoginError).toHaveBeenCalled();
      });
    });
  });
});
