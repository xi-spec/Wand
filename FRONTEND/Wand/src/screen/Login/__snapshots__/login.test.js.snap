// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Login component When is rendered And user.isLogged is false Then is matched to snapshot 1`] = `
<RCTScrollView
  style={
    Object {
      "marginLeft": 50,
      "marginRight": 50,
      "paddingTop": 50,
    }
  }
>
  <View>
    <View>
      <Text
        style={
          Object {
            "color": "#2E3192",
            "fontSize": 36,
            "marginBottom": 5,
          }
        }
      >
        Welcome!
      </Text>
      <Text
        style={
          Object {
            "color": "#979797",
            "fontSize": 18,
          }
        }
      >
        Sign in to continue
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingTop": 50,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        onChangeText={[Function]}
        placeholder="Email"
        placeholderTextColor="#979797"
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomColor": "#979797",
            "borderBottomWidth": 1,
            "color": "#2E3192",
            "fontSize": 18,
            "height": 50,
            "marginBottom": 30,
            "marginTop": 20,
            "width": 300,
          }
        }
        testID="emailInput"
        underlineColorAndroid="transparent"
        value="ho@la.com"
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Password"
        placeholderTextColor="#979797"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "borderBottomColor": "#979797",
            "borderBottomWidth": 1,
            "color": "#2E3192",
            "fontSize": 18,
            "height": 50,
            "marginBottom": 30,
            "marginTop": 20,
            "width": 300,
          }
        }
        testID="passwordInput"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "marginTop": 70,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2E3192",
            "borderRadius": 5,
            "height": 50,
            "justifyContent": "center",
            "marginBottom": 18,
            "opacity": 1,
            "width": 250,
          }
        }
        testID="login"
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
            }
          }
        >
          LOGIN
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingTop": 20,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomColor": "#E6E6E6",
            "borderBottomWidth": 1,
            "marginLeft": 5,
            "marginRight": 5,
            "width": 150,
          }
        }
        underlineColorAndroid="transparent"
      />
      <Text
        style={
          Object {
            "color": "#979797",
            "top": 7,
          }
        }
      >
        o
      </Text>
      <TextInput
        allowFontScaling={true}
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomColor": "#E6E6E6",
            "borderBottomWidth": 1,
            "marginLeft": 5,
            "marginRight": 5,
            "width": 150,
          }
        }
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "opacity": 1,
          "paddingTop": 50,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "https://assets.stickpng.com/images/580b57fcd9996e24bc43c516.png",
          }
        }
        style={
          Object {
            "height": 50,
            "width": 50,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
